endpoints/openrtb2/auction.go

```diff
@@ -157,6 +158,10 @@ func (deps *endpointDeps) validateRequest(req *openrtb.BidRequest) error {
 		return err
 	}
 
+	if err := validateUser(req.User); err != nil {
+		return err
```

Comment:
User is not a required parameter in the BidRequest; do we reject the auction request based on User field validation or can we ignore it and let it pass?

Reply:
@dbemiller what do you think? from our earlier chats I got the sense that we should return an error here.

Reply:
hah! Yes... this is the exact opposite of my comment here: https://github.com/prebid/prebid-server/pull/270#issuecomment-357251853

It's a philosophical question. Given bad user input, which one should software do? The choices are:

1. Fail fast with error messages.
2. Try to infer what the user wants, and do the best thing it can.

If you're unsure, run this thought experiment: suppose we release a buggy version of `prebid.js` which _doesn't_ check the expiration date.

In philosophy (1), PBS fails and returns and error.
In philosophy (2), it runs the auction with invalid user ID info.

Now imagine a publisher who tries to upgrade `prebid.js` versions.

In philosophy (1), they see that PBS returns 400 errors. They probably decide not to upgrade, and if they're good citizens they'll log a ticket about it. It gets fixed, and we release a PBJS patch.

In philosophy (2), PBS auctions _seem_ to work. A month later, they notice that revenue has dropped. After digging painfully through tons of analytics data, they learned that their user ID sync rate has been 0 ever since {some date}. They cross reference that date with changes on their site, and realize that it corresponds with the PBJS release.

Then multiply the cost of this investigation across the entire industry, since every company will be launching their own investigation to see why revenue fell.

so... Fail fast. Fail loud. It's almost always better that way.

Reply:
Was under the impression that more bids => more revenue but I see what you mean. :+1:

Reply:
you're totally right ^^. More bids => more revenue... but longer bugfix investigations => more costs :).